### PR TITLE
Fix comparison of changed files

### LINKED ISSUE #26: New publish concurrent with yanks is missed
From investigation in https://github.com/rust-lang/docs.rs/issues/1912 it looks like this is a `crates-index-diff` issue. The crate had a publish and two yanks in between two checks and we only see the yank event.

The relevant commit range contains just this publish and the two yanks:

```console
> git log --oneline --reverse b49672ff6a2d40123a593cfbca9d05219346e398~1..92c18bdf30a4872d355e4a5b3a7f7c6c75323cf7
b49672ff6a2 Updating crate `aegis#0.2.4`
da97cd0243b Updating crate `ansi-color-codec#0.3.11`
1533f8e863a Yanking crate `ansi-color-codec#0.3.4`
92c18bdf30a Yanking crate `ansi-color-codec#0.3.5`
```

 I wrote a little test program to verify this:

```rust
use crates_index_diff::{Change, CrateVersion};

fn main() -> anyhow::Result<()> {
    let index = crates_index_diff::Index::from_path_or_cloned("index")?;
    let mut args = std::env::args().skip(1);
    let start = git_hash::ObjectId::from_hex(args.next().unwrap().as_bytes())?;
    let end = git_hash::ObjectId::from_hex(args.next().unwrap().as_bytes())?;
    for change in index.changes_between_commits(start, end)? {
        match change {
            Change::Added(CrateVersion { name, version, .. }) => println!("added {name} {version}"),
            Change::Yanked(CrateVersion { name, version, .. }) => println!("yanked {name} {version}"),
            Change::Deleted { name } => println!("deleted {name}"),
        }
    }
    Ok(())
}
```

For the full range it shows the same behaviour:

```console
> cargo run --quiet --release -- b49672ff6a2d40123a593cfbca9d05219346e398 92c18bdf30a4872d355e4a5b3a7f7c6c75323cf7
yanked ansi-color-codec 0.3.4
yanked ansi-color-codec 0.3.5
```

If I only include the first one or two commits it behaves correctly:

```console
> cargo run --quiet --release -- b49672ff6a2d40123a593cfbca9d05219346e398 da97cd0243bd297e7b0e4b11040e531af2e256d1
added ansi-color-codec 0.3.11

> cargo run --quiet --release -- b49672ff6a2d40123a593cfbca9d05219346e398 1533f8e863aee3fd5340513acfeef9d42816cd08
yanked ansi-color-codec 0.3.4
added ansi-color-codec 0.3.11
```

### PR DESCRIPTION
**Pull request by pascalkuthe**
This PR is a major refactor of the core diff logic to fix #26.
The crate completely removes any textual diffing that was previously performed using `imara-diff`.
I verified that `imara-diff` function correctly.
The problem is that `imara-diff` (and diffs in general) are intended to provide linear editing sequences.
That is a set of edits to transform file A into file B **without any jumps**.
However, for `crates-index-diff` we actually don't care about this property, because we only care about how any given
checksum changed (removed/added/yanked).
This means that `imara-diff` provides unideal diffs for this usecase.
For example if a version X is moved from the end of the file to the start of the file imara diff provides to changes:
add X at line 0
remove X at last line

That means that any `remove` changes provided by `imara-diff` had to be ignored for correct functionality.
Ignoring removed lines works when crates are never deleted.
However, if comparing two (far apart) commits between which a crate was deleted and later reclaimed by someone else.
Then we need to mark any removed checksum as deleted.

This PR changes the diff logic to perform a simple unordered `HashMap<Checksum, CrateVersion>` comparison.
The details are a bit tricky to avoid parsing every line (and to avoid storing the checksum twice).
However, the code is functionally equivalent to the following s (non compiling pseudocode, some details left out):

``` rust
let old_versions: HashMap<Checksum, CrateVersion> = diff.old.lines().map(|line| parse_crate_version(line));
let new_versions: HashMap<Checksum, CrateVersion> = diff.new.lines().map(|line| parse_crate_version(line));
for new_version in new_versions.difference(old_versions){
  self.changes.push(Change::Added(new_version))
}
for old_version in old_versions.difference(new_versions){
  self.changes.push(Change::Deleted(old_version))
}
for (old_version, new_version) in old_versions.intersection(new_versions){
  let change = match (old_version.yanked, new_version.yanked) {
      (true, false) => Change::Unyanked(new_version),
      (false, true) => Change::Yanked(new_version),
      _ => continue,
  };
  self.changes.push(change)
}
```

In the process I discovered some ambiguities in the API of `Change` around yanked crates.
To address that I have added two new variants to that enum:

``` rust
Unyanked(CrateVersion)
AddedAndYanked(CrateVersion)
```

This ensures that downstream crates can always tell whether a crate
was yanked/unyanked/added without any extra operations.

Another API change:

Due to the refactor of the diff algorithm, changes are now naturally grouped
by crate. That means additional sorting inside `crates-index-diff` is not required to achieve that property.
To improve performance the sorting by crate-name/version has been removed.
If sorting is required by downstream crates, it is trivial for them to call `changes.sort_by` themselves.
This change is a nice performance boost for any crates that don't care about the order of changes (or require a different sorting then the one that was used by `crates-index-diff`).
Note that the order of the changes is completely non-deterministic (except for being grouped by crate).
So if any downstream crate (or a test inside this crate) requires a deterministic order, it must sort the vector.

This PR fixes the bug discussed in #26 where `crates-index-diff` would report
a different set of changes when splitting the diff across multiple commits.
This PR also fixed the `baseline` test which previously forget to include the `HEAD` commit (and only included ancestors).

### CODE REVIEW

File: src/index/diff/mod.rs [hunk 1]
```diff
@@ -164,6 +164,16 @@ impl Index {
 
     /// Similar to `changes()`, but requires `from` and `to` objects to be provided. They may point
     /// to either `Commit`s or `Tree`s.
+    ///
+    /// # Returns
```
Comment (Byron) [inline src/index/diff/mod.rs:168]: I really love these docs, thank you!

File: src/types.rs [hunk 2]
```diff
@@ -59,6 +71,20 @@ impl Change {
...
+    /// Returns all versions affacted by this change
+    /// The returned slice usually has length 1
+    /// However if a crate was purged from the index by an admin,
+    /// all versions of the pruged crate are returned
+    pub fn versions(&self) -> &[CrateVersion] {
+        match self {
+            Change::Added(v)
+            | Change::Unyanked(v)
+            | Change::AddedAndYanked(v)
+            | Change::Yanked(v) => slice::from_ref(v),
```
Comment (Byron) [inline src/types.rs:88]: `slice::from_ref()` is a little gem that I learned about recently enough to still be excited about it :).

File: tests/baseline.rs [hunk 3]
```diff
@@ -16,35 +18,58 @@ fn all_aggregrated_diffs_equal_latest_version(
...
+
+                    let changes = index.changes_between_commits(old, current)?;
+                    for change in changes {
+                        match change {
+                            Added(v) | AddedAndYanked(v) => {
+                                // found a new crate, add it to the index
+                                versions.insert(v.checksum.to_owned());
+                            }
+                            Unyanked(v) | Yanked(v) => {
+                                // yanked/unyanked crates must be part of the index
```
Comment (Byron) [inline tests/baseline.rs]: Loving this added assertion for extra sanity checking :).

**Review by Byron — CHANGES_REQUESTED** [review]
Fantastic work, thanks so much!

Thanks to `git-lfs` I can't push my changes to the PR directly, but have attached the patches here.

[patches.zip](https://github.com/Byron/crates-index-diff-rs/files/10050659/patches.zip) (_Updated_)


Before merging, I'd love if you could try to remove the `unsafe` from the code as it's unclear if these carry any weight without profiling a before and after. From the profile attached here (diffing commit by commit) the time for the actual diffing is marginal at best, and even though bigger steps (i.e. more changes) will result in more work for the differ, I doubt it can stand up to the time it takes to decompress objects. `docs.rs` pulls regularly and there is only ~650 changes per day.

Screenshot-2022-11-20-at-20 27 12

And with that it should be ready to merge.

File: src/types.rs (hunk 2 above)
Comment (pascalkuthe) [reply src/types.rs:88]: I love this function! It's perfect for enapsulating `one or many` without another enum.
I found this little gem in the cranelift code, at some point

**Review by Byron — APPROVED** [review]
And it's a wrap! Thanks so much for all the effort that you put into this project, it's on an entirely new level thanks to it and I think it might finally be 'done' (despite #30 which might be implemented one fine day or maybe never).

**Pull request merged**